src/Http/Http.Abstractions/src/Routing/RouteValueDictionary.cs

```diff
@@ -753,12 +750,8 @@ public PropertyStorage(object value)
 
             // Cache the properties so we can know if we've already validated them for duplicates.
             var type = Value.GetType();
-            if (!_propertyCache.TryGetValue(type, out Properties!))
-            {
-                Properties = PropertyHelper.GetVisibleProperties(type);
-                ValidatePropertyNames(type, Properties);
-                _propertyCache.TryAdd(type, Properties);
-            }
+            Properties = PropertyHelper.GetVisibleProperties(type);
+            ValidatePropertyNames(type, Properties);
```

Comment:
The goal of the cache on this type is to avoid executing ValidatePropertyNames multiple times. See comment earlier in method.

Reply:
But `PropertyHelpers.GetVisibleProperties already has its own cache to avoid doing reflection every time it is invoked. https://github.com/dotnet/aspnetcore/blob/6ac0656f5ca09f035de1c942d20462394c5266f7/src/Shared/PropertyHelper/PropertyHelper.cs#L421-L432.

I will remove the comment.

Reply:
If you look a bit below in `ValidatePropertyNames`, it's looping over the full set and allocating every call - that's what James is talking about avoiding in the current code here. We'd be saving a bit of memory cache on the prop, but adding allocations (eating more net memory) and increasing CPU as a tradeoff.

Reply:
Oh, that's right, hadn't seen them. I add the `ConditionalWeakTable` again.

Reply:
Doesn't a `ConditionalWeakTable` defeat the purpose of the cache? It looks like the caches will be dropped as soon as there aren't any active requests referencing a `RouteValueDictionary` for a given type.

I saw this suggestion in the issue from @davidfowl.

> I could imagine a universe where we do some work here to avoid this cache if we're not in the default load context and avoid this code path for internal usage (if we don't already).

Could we do this instead?

Reply:
How about updating GetVisibleProperties to check for duplicates and cache that along with the properties?

`public static (Properties: PropertyHelper[], HasDuplicates: bool) GetVisibleProperties(...)`

(I'd prefer a class instead of a tuple for the return type, but you get the idea)

Reply:
Or we could remove the caching from PropertyHelper and cache it just here. `PropertyHelper.GetVisiblePropertiesUncached` for e.g.. `PropertyHelper` does a bunch of things (it's a discoverer / factory, a cache. it's derivable for some reason) and I feel it might be better for it be good at one thing (discovery) and have the callers cache results as necessary.